### PR TITLE
test: add e2e test for rootless mode

### LINKED ISSUE #503: Improve e2e testing
Currently we deploy vcluster in a single configuration in parallel in 4 different jobs for each distro. While this works and smoke tests vcluster, several features and areas of vcluster remain untested due to this single deployment of vcluster. Hence I propose to deploy vcluster within those jobs multiple times in different configurations, so that we can test certain features as well.

Currently untested vcluster functionality (might be incomplete):
- [ ] default installation with namespace admin RBAC only
- [x] rootless mode
- [x] enable scheduler
- [x] sync all nodes
- [ ] vcluster in vcluster
- [x] sync nodes based on node selector
- [x] isolated mode
- [x] target namespace
- [x] pause / resume vcluster
- [ ] persistentvolumes
- [ ] storageclasses
- [ ] priorityclasses
- [ ] serviceaccounts
- [ ] plugins

### PR DESCRIPTION
**Pull request by pratikjagrut**
**What issue type does this pull request address?** (keep at least one, remove the others)
/kind test

**What does this pull request do? Which issues does it resolve?** (use `resolves #<issue_number>` if possible) 
resolves #503

### CODE REVIEW

File: test/e2e_rootless/values.yaml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+securityContext:
+  runAsUser: 12345
+  runAsNonRoot: true
+
```
Comment (olljanat) [inline test/e2e_rootless/values.yaml:4]: assuming that #566 will get merged you should also add:
```yaml
fsGroup: 12345
```

File: test/e2e_rootless/values.yaml [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+## this values are for k8s distro
+syncer:
+  securityContext:
+    runAsUser: 12345
+    runAsNonRoot: true
```
Comment (olljanat) [inline test/e2e_rootless/values.yaml:9]: assuming that #566 will get merged you should also add:
```yaml
etcd:
  fsGroup: 12345
```

**Review by olljanat — COMMENTED** [review]
Thanks for this PR. We have been running K8s as rootless longtime already without issues. Only exception is that etcd needs fsGroup value which our policy controller adds automatically. Created PR #566 to fix it correctly in Helm charts now. 

It is also possible to harden those pods even more which why we most probably should test least privileged model at least with some variant. 

Here is maximum hardening which I have managed to get working. With those securityContexts vcluster will be able to run even on host cluster where https://kubernetes.io/docs/concepts/security/pod-security-standards/#restricted is enforced:
```yaml
syncer:
  securityContext:
    allowPrivilegeEscalation: false
    capabilities:
      drop:
      - ALL
    readOnlyRootFilesystem: true
    runAsGroup: 12345
    runAsNonRoot: true
    runAsUser: 12345
    seccompProfile:
      type: RuntimeDefault

etcd:
  fsGroup: 12345
  securityContext:
    allowPrivilegeEscalation: false
    capabilities:
      drop:
      - ALL
    readOnlyRootFilesystem: true
    runAsGroup: 12345
    runAsNonRoot: true
    runAsUser: 12345
    seccompProfile:
      type: RuntimeDefault

controller:
  securityContext:
    allowPrivilegeEscalation: false
    capabilities:
      drop:
      - ALL
    readOnlyRootFilesystem: true
    runAsGroup: 12345
    runAsNonRoot: true
    runAsUser: 12345
    seccompProfile:
      type: RuntimeDefault

api:
  securityContext:
    allowPrivilegeEscalation: false
    capabilities:
      drop:
      - ALL
      add:
      - NET_BIND_SERVICE
    readOnlyRootFilesystem: true
    runAsGroup: 12345
    runAsNonRoot: true
    runAsUser: 12345
    seccompProfile:
      type: RuntimeDefault
```
But when tested it I found that seccompProfile is actually missing from pre-install job and CoreDNS which user cannot configure so created #567 to fix that issue.

However if these are too much for you PR I can created another one later to modify those rootless tests.

**Comment by pratikjagrut** [comment]
> However if these are too much for you PR I can created another one later to modify those rootless tests.

@olljanat Maybe we could have separate e2e for this because your config also has hardened security and isolation mode.

File: .github/workflows/e2e.yaml [hunk 3]
```diff
@@ -171,12 +175,18 @@ jobs:
...
       run: |
         set -x 
+        kubectl get pods -o yaml -n ${{ env.VCLUSTER_NAMESPACE }}
+        echo "======================================================================================================================"
         kubectl get events -n ${{ env.VCLUSTER_NAMESPACE }} --sort-by='.lastTimestamp'
         echo "======================================================================================================================"
-        kubectl logs -l app=${{ env.VCLUSTER_SUFFIX }} -n ${{ env.VCLUSTER_NAMESPACE }} -c syncer --tail=5000
+        kubectl logs -l app=${{ env.VCLUSTER_SUFFIX }} -n ${{ env.VCLUSTER_NAMESPACE }} -c syncer --tail=500
+        echo "======================================================================================================================"
+        kubectl logs -l app=${{ env.VCLUSTER_SUFFIX }} -n ${{ env.VCLUSTER_NAMESPACE }} -c vcluster --tail 500
```
Comment (matskiv) [inline .github/workflows/e2e.yaml:188]: Same as above

File: test/e2e_rootless/rootlessmode.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+				Container: "syncer",
+				Command:   cmd,
+				Stdout:    true,
+			},
+			clientgoscheme.ParameterCodec,
+		)
+		exec, err := remotecommand.NewSPDYExecutor(f.HostConfig, "POST", req.URL())
+		framework.ExpectNoError(err)
+		var stdout bytes.Buffer
+		err = exec.Stream(remotecommand.StreamOptions{Stdout: &stdout})
```
Comment (matskiv) [inline test/e2e_rootless/rootlessmode.go:40]: btw, we have util to make exec easier - https://github.com/loft-sh/vcluster/blob/main/pkg/util/podhelper/exec.go#L152-L153

**Review by matskiv — CHANGES_REQUESTED** [review]
A change in .github/workflows/e2e.yaml is needed IMHO
The other comment is FYI only - change optional :)

**Review by matskiv — APPROVED** [review]
LGTM. Thanks! :+1:

**Review by FabianKramm — APPROVED** [review]
@pratikjagrut thanks for the PR! LGTM!

**Pull request merged**